### DISCUSSION TITLE
What's the best way to check if an `openai.Client()` was patched by instructor?

### DISCUSSION THREAD

**Discussion by gustavocidornelas** [567-labs/instructor#457] [Q&A] [answered]
I want to know if there's a good way to programmatically check if an `openai.Client()` has been patched by instructor or if it is the "original" OpenAI client. 

For example:
```python
import instructor
import openai

openai_client = openai.Client()
# openai_client is the original OpenAI client -> so its chat.completions.create method will return a `ChatCompletion` object

patched_client = instructor.patch(openai_client)
# patched_client was patched by instructor -> so its chat.completions.create method might return an object defined by `response_model`

```
I'd like to know what's the best way to write a function like `is_patched_client(client: openai.Client) -> bool`, such that:
```python
is_patched_client(openai_client)
# returns False

is_patched_client(patched_client)
# returns True
```
I can obviously check the type of the object I get from `chat.completions.create`, but ideally, I'd like to know if the client was patched or not just based on the client itself.

Any help is appreciated!

**Comment by jxnl** [comment, marked as answer]
this will be possible in 1.0 if you can wait a week or so! It'll be 

`type(client) == InstructorOpenAI`

**Reply by gustavocidornelas** [reply]
kk, thanks! In the meantime, I'll keep checking the output type.

Also, I'd be great to see some instructor-related metadata as part of the response model, such as the number of retries that actually happened, the errors (stack traces), etc. I'm just thinking out loud now, I haven't checked if that's already possible or not :)

**Reply by jxnl** [reply]
we do that as part of logging

where would you want to see it all? like 

```
resp = create()
resp._metadata? 
```

**Reply by gustavocidornelas** [reply]
Yes, something like that. As part of `resp`, continue to be able to access `resp._raw_response` and additionally have `resp._metadata` where I could see some instructor-specific metadata.

**Reply by gustavocidornelas** [reply]
The way I'm thinking about it is:
- If I have a lot of retries, it means that the model is having a hard time conforming to the structure I expect. 
- The stack trace tells me specifically which fields from the response model were particularly challenging. 
- Given my interest in minimizing the number of retries to keep the system's response time low, the stack traces tell me exactly where my prompt needs polishing (either via Pydantic or via the regular `messages` prompt).
- Looking at the logs is not ideal if I want to do this at scale. If I can access this metadata as part of the response, I can monitor and test it.

**Reply by jxnl** [reply]
sounds good, we'd have to keep it as BaseModel

since 

`response_model=bool` will be impossible to attach data too without giving you some wrapper hell.
